### PR TITLE
KAFKA-12648: fixes for query APIs with named topologies

### PR DESCRIPTION
**Pull request by ableegoldman**
Fixes some issues with the NamedTopology version of the IQ methods that accept a `topologyName` argument, and adds tests for all.

### CODE REVIEW

**Comment by guozhangwang** [comment]
@ableegoldman some related tests are failing, I will go ahead and review the code still.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java [hunk 1]
```diff
@@ -296,9 +297,10 @@ private boolean hasPartitionsForAnyTopics(final List<String> topicNames, final S
     }
 
     private Set<String> getStoresOnHost(final Map<String, List<String>> storeToSourceTopics,
-        final Set<TopicPartition> sourceTopicPartitions, final String topologyName) {
+                                        final Set<TopicPartition> sourceTopicPartitions,
+                                        final String topologyName) {
         final InternalTopologyBuilder builder = topologyMetadata.lookupBuilderForNamedTopology(topologyName);
-        final Set<String> sourceTopicNames = builder.sourceTopicNames();
+        final Collection<String> sourceTopicNames = builder.sourceTopicCollection();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: Do we want the raw topic names (without the prefix) or the decorated ones here? 

BTW The function/variable names are a bit confusing but they stored different things. Maybe we should just rename them to be more clear.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java [hunk 2]
```diff
@@ -349,8 +351,10 @@ private void rebuildMetadata(final Map<HostInfo, Set<TopicPartition>> activePart
                     final Map<String, Collection<String>> namedTopologyToStoreName = new HashMap<>();
                     final Set<String> topologyNames = topologyMetadata.namedTopologiesView();
                     topologyNames.forEach(topologyName -> {
-                        final Collection<String> storesOnHostForTopologyName = getStoresOnHost(storeToSourceTopics, activePartitionHostMap.get(hostInfo), topologyName);
-                        storesOnHostForTopologyName.addAll(getStoresOnHost(storeToSourceTopics, standbyPartitionHostMap.get(hostInfo), topologyName));
+                        final Map<String, List<String>> topologyStoresToSourceTopics =
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: What's the difference between `topologyStoresToSourceTopics` and `storeToSourceTopics` here?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: We want the actual ("decorated") topic names. PR #11562 introduced a bug by using the `sourceTopicNames` with only the raw topic names, this is one of the main IQ fixes. 

100% agree that the names are confusing, I was going to address that in #11601 but I suppose it makes sense to do that in this PR as well

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: Ack! Thanks.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java [hunk 3]
```diff
@@ -113,7 +113,7 @@ public StreamsMetadata getLocalMetadata() {
             return allMetadata;
         }
 
-        final Collection<String> sourceTopics = topologyMetadata.sourceTopicsForStore(storeName);
+        final Collection<String> sourceTopics = topologyMetadata.sourceTopicsForStore(storeName, null);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:126]: Before this change, `StreamsMetadataState#getAllMetadataForStore(string)` returns the metadata for the store names across all topologies (it's possible that multiple topologies have stores with the same name), whereas now it is only returning for the `UNNAMED_TOPOLOGY`, is this change intentional?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java [hunk 4]
```diff
@@ -428,7 +469,7 @@ private void rebuildMetadata(final Map<HostInfo, Set<TopicPartition>> activePart
         final Set<HostInfo> standbyHosts = new HashSet<>();
         for (final StreamsMetadata streamsMetadata : allMetadata) {
             if (streamsMetadata instanceof NamedTopologyStreamsMetadataImpl
-                && ((NamedTopologyStreamsMetadataImpl) streamsMetadata).namedTopologyToStoreNames().get(topologyName).contains(storeName)) {
+                && ((NamedTopologyStreamsMetadataImpl) streamsMetadata).storeNamesByTopology().get(topologyName).contains(storeName)) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: Hmm.. this seems not right to me. Below we want to collect both active and standby host infos in the `KeyQueryMetadata`, but here we would only enter if the active store names contains it.

If that should be fixed, then it seems that `storeNamesByTopology()` and `standbyStoreNamesByTopology()` always appears hand-in-hand, hence I'm wondering if it's necessary to keep them in two fields and two getter functions?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: This comment is the lines 473 - 476 below: should we use the `topologyName` as well? Otherwise, e.g. 

```
final Set<String> standbyStateStoreNames = streamsMetadata.standbyStateStoreNames();
```

Would be getting for the `UNNAMED` topology, which seems not correct?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java [hunk 5]
```diff
@@ -296,9 +316,10 @@ private boolean hasPartitionsForAnyTopics(final List<String> topicNames, final S
     }
 
     private Set<String> getStoresOnHost(final Map<String, List<String>> storeToSourceTopics,
-        final Set<TopicPartition> sourceTopicPartitions, final String topologyName) {
+                                        final Set<TopicPartition> sourceTopicPartitions,
+                                        final String topologyName) {
         final InternalTopologyBuilder builder = topologyMetadata.lookupBuilderForNamedTopology(topologyName);
-        final Set<String> sourceTopicNames = builder.sourceTopicNames();
+        final Collection<String> sourceTopicNames = builder.fullSourceTopicNames();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: Just to clarify this is the actual fix right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java [hunk 6]
```diff
@@ -324,6 +345,45 @@ private void rebuildMetadata(final Map<HostInfo, Set<TopicPartition>> activePart
...
+            .distinct()
+            .forEach(hostInfo -> {
+                final Set<TopicPartition> activePartitionsOnHost = new HashSet<>();
+                final Set<TopicPartition> standbyPartitionsOnHost = new HashSet<>();
+
+                final Map<String, Set<String>> storeNamesByTopology = new HashMap<>();
+                final Map<String, Set<String>> standbyStoreNamesByTopology = new HashMap<>();
+
+                topologyMetadata.namedTopologiesView().forEach(topologyName -> {
+                    // WIP
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: ?? Still WIP?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java [hunk 7]
```diff
@@ -324,6 +345,45 @@ private void rebuildMetadata(final Map<HostInfo, Set<TopicPartition>> activePart
...
+                final Map<String, Set<String>> storeNamesByTopology = new HashMap<>();
+                final Map<String, Set<String>> standbyStoreNamesByTopology = new HashMap<>();
+
+                topologyMetadata.namedTopologiesView().forEach(topologyName -> {
+                    // WIP 
+                });
+
+                final NamedTopologyStreamsMetadataImpl metadata = new NamedTopologyStreamsMetadataImpl(
+                    hostInfo,
+                    storeNamesByTopology,
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: Since this PR is not completed yet (btw could you rename it to indicate it's WIP?), I'm not sure what's the motivation from changing it to a mapping from topology to names, but I feel we may not need to maintain two mappings for active and standby since we are always trying to check if store name is in either of them, is that right?

**Review by guozhangwang — COMMENTED** [review]
Thanks @ableegoldman , please ping me when the PR is ready for reviews.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java (hunk 3 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:126]: Yes, the reason is that a single store name is not sufficient to identify the actual store with named topologies, since a name may be reused for different stores in different topologies. So a user must use the overload that also accepts a `topologyName` parameter.

I will add a check that throws an informative exception if this method is called when named topologies are used, to be safe.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java (hunk 5 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: One of them, yes

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java (hunk 4 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: See the latest commit, I refactored how we handle this so that each StreamsMetadata object is associated with a single `topologyName`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java (hunk 6 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: Ready for review now, sorry about that!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java (hunk 7 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java]: See the latest changes -- lmk if you still have a question here

**Comment by ableegoldman** [comment]
Hey @guozhangwang , thanks for the initial review and sorry for the delay, I've been sitting on this PR while I worked on the tests but should have pushed the non-testing code in the meantime. Anyways I've pushed the final iteration on this, redid the overall approach a bit and think it makes a lot more sense now. Lmk what you think, and I'll let you know when the testing code is done/ready for review

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 8]
```diff
@@ -1140,6 +1140,11 @@ private void completeShutdown(final boolean cleanRun) {
 
         topologyMetadata.unregisterThread(threadMetadata.threadName());
 
+        try {
+            topologyMetadata.unregisterThread(threadMetadata.threadName());
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1142]: You might need to change line 1141?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java [hunk 9]
```diff
@@ -224,6 +225,7 @@ public void maybeWaitForNonEmptyTopology(final Supplier<StreamThread.State> thre
         } finally {
             unlock();
         }
+        log.info("Removed NamedTopology {} and updated topology version to {}", topologyName, version.topologyVersion.get());
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: line 220 has the same log

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java [hunk 10]
```diff
@@ -138,6 +138,7 @@ public void registerThread(final String threadName) {
 
     public void unregisterThread(final String threadName) {
         threadVersions.remove(threadName);
+        maybeNotifyTopologyVersionWaitersAndUpdateThreadsTopologyVersion(threadName);
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:141]: Do we really need this here?

**Review by wcarlson5 — COMMENTED** [review]
just a few small things. Also you have a few test failures that look related i.e. 
```
org.apache.kafka.streams.integration.StoreQueryIntegrationTest.shouldQuerySpecificStalePartitionStoresMultiStreamThreadsNamedTopology
```
Can you clean those up?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 8 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1142]: Oh good catch, I guess that got missed since I ported this fix from [#11601](https://github.com/apache/kafka/pull/11601/) (since that doesn't need to make 0.24) -- fixed

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java (hunk 10 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:141]: Yeah, sorry I should have left an explanation for this -- basically if we have `#add/removeNamedTopology` futures blocked on this one thread ack'ing the topology update, and that update occurs between when the thread checks for these updates and when it checks for/goes into shutdown, then we need to update the waiters and potentially complete any that have just been waiting on this thread which has missed the update before shutting down

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java (hunk 9 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: There's a semantic difference but it's hard to notice, the first one says "Remov_ing_" when we start removing it, whereas this one says "Remov_ed_" to indicate that we have finished. That said it's probably worth clarifying further, eg `"Finished removing..."` -- I'll update it to make the purpose of each line more clear

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:126]: Got it, thanks for the notes.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/StreamsMetadataImpl.java [hunk 11]
```diff
@@ -153,12 +166,19 @@ public int hashCode() {
...
                 ", topicPartitions=" + topicPartitions +
                 ", standbyStateStoreNames=" + standbyStateStoreNames +
                 ", standbyTopicPartitions=" + standbyTopicPartitions +
-                '}';
+            '}';
+        if (topologyName == null) {
+            return str;
+        } else {
+            return str +
+                "topologyName=" + topologyName;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/StreamsMetadataImpl.java]: nit: add a space so it is "StreamsMetadata {...} topologyName=xyz"

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java [hunk 12]
```diff
@@ -319,6 +315,14 @@ private void verifyTopologyStateStore(final String topologyName, final String st
         return streamsMetadataState.getAllMetadataForStore(storeName, topologyName);
     }
 
+    /**
+     * See {@link KafkaStreams#metadataForAllStreamsClients()}
+     */
+    public Collection<StreamsMetadata> allStreamsClientsMetadataForTopology(final String topologyName) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java:321]: Not needed for this PR, but I see there's an inconsistency in the function names: these two e.g. adds a `ForTopology` suffix to differentiate with the original ones in `KafkaStreams`, but others just overload the same function. When we expose them in the KIP we need to make it consistent.

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java [hunk 13]
```diff
@@ -319,23 +329,88 @@ public void shouldProcessMultipleIdenticalNamedTopologiesWithInMemoryAndPersiste
...
+            streams2.start(asList(topology1Builder2.build(), topology2Builder2.build()));
+            waitForApplicationState(asList(streams, streams2), State.RUNNING, Duration.ofSeconds(30));
+
+            final Collection<StreamsMetadata> streamsMetadataForStoreTopology1 = streams.streamsMetadataForStore("store", TOPOLOGY_1);
+            final Collection<StreamsMetadata> streamsMetadataForStoreTopology2 = streams.streamsMetadataForStore("store", TOPOLOGY_2);
+            final Collection<StreamsMetadata> streams2MetadataForStoreTopology1 = streams2.streamsMetadataForStore("store", TOPOLOGY_1);
+            final Collection<StreamsMetadata> streams2MetadataForStoreTopology2 = streams2.streamsMetadataForStore("store", TOPOLOGY_2);
+
+            // we don't know which host was actually assigned the task  containing the state
+            // tore in topology-2, only that there should be exactly one such host
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java]: nit: `store`

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java [hunk 14]
```diff
@@ -319,23 +329,88 @@ public void shouldProcessMultipleIdenticalNamedTopologiesWithInMemoryAndPersiste
...
-    public void shouldAddNamedTopologyToUnstartedApplicationWithEmptyInitialTopology() throws Exception {
-        topology1Builder.stream(INPUT_STREAM_1).groupBy((k, v) -> k).count(IN_MEMORY_STORE).toStream().to(OUTPUT_STREAM_1);
-        streams.addNamedTopology(topology1Builder.build());
-        IntegrationTestUtils.startApplicationAndWaitUntilRunning(singletonList(streams), Duration.ofSeconds(15));
+    public void shouldAddNamedTopologiesBeforeStartingAndRouteQueriesToCorrectTopology() throws Exception {
+        try {
+            // for this test we have one of the topologies read from an input topic with just one partition so
+            // that there's only one instance of that topology's store and thus should always have exactly one
+            // StreamsMetadata returned by any of the methods that look up all hosts with a specific store and topology
+            CLUSTER.createTopic(SINGLE_PARTITION_INPUT_STREAM, 1, 1);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:340]: Nice coverage with different num.partitions, thanks!

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java [hunk 15]
```diff
@@ -635,6 +710,19 @@ public void shouldWaitForMissingInputTopicsToBeCreated() throws Exception {
         }
     }
 
+    /**
+     * @return  true iff the metadata contents of each StreamsMetadata object are equal, ie all fields except
+     *          for the topologyName
+     */
+    private static boolean metadataIsEqual(final StreamsMetadata left, final StreamsMetadata right) {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java]: Should we check topology name from casting to Impl as well?

**Review by guozhangwang — COMMENTED** [review]
Made a final path.

**Comment by guozhangwang** [comment]
LGTM! Jenkins failed on compilation (checkstyle). Please feel free to merge after addressing the nit comments and green build.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java (hunk 12 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java:321]: Fair, I was thinking of them as semantically different since for example in this case we are changing the meaning of the original method to return only a subset of metadata. Whereas for something like `#store`, I felt it made sense to keep the same name since the meaning of this API doesn't change when using named topologies -- you just happen to need one additional parameter in order to uniquely identify a state store since the name alone could point to multiple stores from different topologies. But it's definitely something to discuss when we get to the KIP proposal

**Comment by ableegoldman** [comment]
test failures are unrelated, ready to merge
